### ISSUE TITLE
several issues

### ISSUE THREAD

**Issue by HuangBo-Terraloupe** [fizyr/keras-retinanet#157]
First of all, the network works really good, I did experiment by using faster rcnn and retina-net, retina-net work much better.
 
![regensburg006354](https://user-images.githubusercontent.com/26878250/33618612-4ea3a9e0-d9e3-11e7-869d-f5b05161c94a.jpg)

I am confused with several things compare to the old faster rcnn repo.

1, I found these in retinanet.py:   sizes   = [32, 64, 128, 256, 512],  strides = [8, 16, 32, 64, 128]
    I got 2 questions for this: 
    a, can we modify the number of these anchors? If we have a anchor size 1024, we also should modify the stride to 256?
    b, Should we wrap or reshape the image to match the anchor? I mean in faster rcnn, it will wrap the image, if the image size is less than 600. I also see this operation in the testing code for coco-dataset.
My feelings is, actually 

2, I did not see the image normalization for csv training(maybe I did not find it), when I do testing, I also did not normalise the image before send it to the network. But it works, object still can be detected.
So my question is: Should we do image normalization for csv training and testing?

**Comment by hgaiser** [comment]
> First of all, the network works really good, I did experiment by using faster rcnn and retina-net, retina-net work much better.

Cool! Thanks for sharing your results, looks good :)

> a, can we modify the number of these anchors? If we have a anchor size 1024, we also should modify the stride to 256?

strides is actually something that is derived from the network backbone. As they describe in the retinanet paper, they use ResNet and append a sixth and seventh stage at the end, resulting in stages 3, 4, 5, 6 and 7 being used for object detection. For stage 3, the image has been downsampled 8 times, stage 4 has been downsampled 16 times and so on. I'm not sure if it is true, but I think the sizes relate to the receptive field of each pixel in a certain stage. In other words, one "pixel" in the third stage is influenced by an 32x32 area in the original image, one "pixel" in the fourth stage by a 64x64 area and so on.

The number of sizes/strides must match the number of pyramids, so you cannot simply add an entry to those lists without also adding a pyramid level (the stages in the text above).

Alternatively you can modify the number if scales / ratios that are used for each "pixel" in each stage. Currently there are 3 scales and 3 ratios, meaning a total of 9 anchors are being generated per "pixel" in each stage. I suppose, judging by your example, your objects are always more or less square, meaning you could for instance generate 6 scales and only 1 ratio. The current version of `keras-retinanet` falls a bit short in allowing these values to be easily changed, but it's something I'm hoping to look into after the basic functionality (read: replicate the paper results as best as possible) is finished.

> b, Should we wrap or reshape the image to match the anchor? I mean in faster rcnn, it will wrap the image, if the image size is less than 600. I also see this operation in the testing code for coco-dataset.

I'm not sure what you mean here. The current generators always reshape the images such that the shortest sides is 600px, unless the larger side then becomes larger than 1024 (in that case it rescales such that the largest side is 1024). This isn't strictly speaking necessary in `keras-retinanet` since it is a fully convolutional network, but it puts an upper bound on the memory required. Nothing is theoretically stopping you from processing arbitrarily shaped images.

> My feelings is, actually
> 
> 2, I did not see the image normalization for csv training(maybe I did not find it), when I do testing, I also did not normalise the image before send it to the network. But it works, object still can be detected.
> So my question is: Should we do image normalization for csv training and testing?

The only form of normalization currently used is that of [subtracting the ImageNet mean](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/utils/image.py#L30). The reason for this is that it is the only preprocessing done for the pretrained resnet weights on imagenet.

This being said, this isn't really an issue but a discussion on the internals of `keras-retinanet`. I'd be happy to discuss further on the Slack channel we have for topics like this (see the `README.md` for instructions). I'll go ahead and close this issue.

**Comment by HuangBo-Terraloupe** [comment]
Sorry, I did not make my point clear, first of all, all the things I am talking about is only related to training on a custom dataset.

1, You mean the csv generater will reshape the image if the image size is shorter than 600 for training? If it does so, I should also first reshape the image to 600, then run the testing.
2, For csv training, you mean so far we normalize image using ImageNet mean? 
If so, I will try to contribute a code for normalize the image using custom image mean, if we will train the model on a custom dataset.

This is what I confused, I did not find the reshape code and normalization code for csv training. Because I should do the same thing for testing. What strange is, the testing actually works.

**Comment by hgaiser** [comment]
> 1, You mean the csv generater will reshape the image if the image size is shorter than 600 for training? If it does so, I should also first reshape the image to 600, then run the testing.

Yes the CSV generator will reshape such that the smallest side is 600 (you can adjust this by changing `image_min_side` and `image_max_side` in the constructor of the `CSVGenerator`).

> 2, For csv training, you mean so far we normalize image using ImageNet mean?
> If so, I will try to contribute a code for normalize the image using custom image mean, if we will train the model on a custom dataset.

Yes so far we normalize by imagenet mean. I'm pretty sure a custom image mean would not make a difference, unless perhaps your images are in a completely different range (say they're 16 bit instead of 8 bit per channel).

> This is what I confused, I did not find the reshape code and normalization code for csv training. Because I should do the same thing for testing. What strange is, the testing actually works.

That it still works if you don't reshape isn't that odd, since it should be relatively size invariant. Not subtracting the imagenet mean could have more severe impact.

ps. the code you're looking for is "hidden" in the superclass, [here](https://github.com/fizyr/keras-retinanet/blob/master/keras_retinanet/preprocessing/generator.py#L109-L127).
